### PR TITLE
Singleton class

### PR DESCRIPTION
**Pull request by rajeevgUCI**
Created Singleton class and refactored GameManager to use it.

Testing: wrote tests for Singleton class. Removed GameManagerTests that were only testing singleton functionality.

### CODE REVIEW

**Comment by rajeevgUCI** [comment]
(Rewrite still in progress, do not review yet.)

File: Assets/Runtime/core/GameManager.cs [hunk 1]
```diff
@@ -1,18 +1,11 @@
...
 
   private void Awake() {
-    if (manager == null) {
-      GameManager.Manager = this;
-      DontDestroyOnLoad(manager);
-      this.Initialize();
-    } else if (manager != this) {
-      Destroy(this.gameObject); // Delete current instance to maintain singleton.
-    }
+    this.Initialize();
```
Comment (kylebartz) [inline Assets/Runtime/core/GameManager.cs]: The null check safety around Initialize is necessary here. GameManager's Initialize should only ever be called once no matter what. Without a safety check, in the off chance that there are multiple GameManagers in the scene there will be unintended effects (such as double bindings being registered). You can quickly test this by adding two game managers on the scene on master branch vs this branch.

File: Assets/Runtime/core/GameManager.cs [hunk 2]
```diff
@@ -1,18 +1,11 @@
 ﻿using System;
 using UnityEngine;
 
-public class GameManager : MonoBehaviour {
-  private static GameManager manager = null;
+public class GameManager : SingletonMonoBehavior<GameManager> {
   private GameState state = GameState.ACTIVE;
 
   private void Awake() {
```
Comment (kylebartz) [inline Assets/Runtime/core/GameManager.cs]: This Awake method is overriding the inherited `Awake()` from `SingletonMonoBehavior`. That means your instance check and deleting if duplicate will never run. It is then possible for multiple instances of GameManger to exist and be active in the scene. Only the first will be registered as the singleton instance, but the rest will still be active and ticking on Update

One solution would be to set `SingletonMonoBehavior`'s `Awake` method to `virtual`, and then have `GameManager`'s `override` and call the base Awake.

File: Assets/Tests/Runtime/core/SingletonTests.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
+﻿using UnityEngine;
+using UnityEngine.TestTools;
+using NUnit.Framework;
+using System.Collections;
+
+public class SingletonTests {
```
Comment (kylebartz) [inline Assets/Tests/Runtime/core/SingletonTests.cs:6]: This is way over commented to the point where it is hard to read the code. In the future try to keep comments minimal.

**Review by kylebartz — CHANGES_REQUESTED** [review]

File: Assets/Runtime/core/GameManager.cs (hunk 1 above)
Comment (rajeevgUCI) [reply Assets/Runtime/core/GameManager.cs]: Good catch - will fix.

File: Assets/Runtime/core/GameManager.cs (hunk 2 above)
Comment (rajeevgUCI) [reply Assets/Runtime/core/GameManager.cs]: I missed that Unity private methods inherit, unlike normal C# private methods. Will fix per your suggestion, then make sure Unity won't e.g. ignore the override and call both Awake methods.

File: Assets/Tests/Runtime/core/SingletonTests.cs (hunk 3 above)
Comment (rajeevgUCI) [reply Assets/Tests/Runtime/core/SingletonTests.cs:6]: The comments document subtleties not apparent from reading the code. So I think it's best to keep their content, but I'll try to make them more concise.

File: Assets/Runtime/core/GameManager.cs (hunk 2 above)
Comment (rajeevgUCI) [reply Assets/Runtime/core/GameManager.cs]: Fixed. Also similarly fixed OnDestroy.

File: Assets/Runtime/core/GameManager.cs (hunk 1 above)
Comment (rajeevgUCI) [reply Assets/Runtime/core/GameManager.cs]: Fixed.

**Comment by rajeevgUCI** [comment]
Made changes per feedback above. Currently resolving merge conflicts.

**Review by kylebartz — APPROVED** [review]

**Pull request merged**